### PR TITLE
[backport] Fix in-place update of arrays in `Link.serialize` and `optimizers.Adam`

### PR DESCRIPTION
**Pull request by emcastillo**
Backport of #7918

### CODE REVIEW

**Comment by emcastillo** [comment]
[automatic post] Jenkins, test this please.

File: tests/chainer_tests/test_link.py [hunk 1]
```diff
@@ -2466,4 +2649,32 @@ def test_to_intel64(self):
         assert l.child.to_method_called == 1
 
 
+@_inject_backend_tests
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: This test did not exist in the previous PR.

Comment (emcastillo) [reply tests/chainer_tests/test_link.py]: Ok!

**Comment by emcastillo** [comment]
PTAL.
Some of the optimizers were not explicitly implemented in the stable version so they had to be skipped too

**Comment by niboshi** [comment]
Jenkins, test this please

**Pull request merged**